src/hotkey/queue.c

```diff
@@ -0,0 +1,190 @@
+#include "queue.h"
+#include "constant.h"
+
+#include <cc_debug.h>
+#include <cc_mm.h>
+#include <cc_pool.h>
+
+#define QUEUE_MODULE_NAME "hotkey::queue"
+
+static bool queue_init = false;
+static int queue_size = 0;
+
+FREEPOOL(qn_pool, qnq, queue_node);
+static struct qn_pool qnp;
+static bool qnp_init = false;
+
+struct queue_node {
+    char                     key[MAX_KEY_LEN];
```

Comment:
One thought- most of the storage used here are for storing keys, and we are storing it (at most) twice, once in the queue/window, once in the key-counter map. However, one property of the data structures is that every key in the map will be in the queue at at least one location. Therefore, it should be possible to use a pointer in the hash table/map to point to the key as stored in the queue. True it's not too big a deal (~ a few MBs) given the number of samples we stored, but I kinda like the idea.